monai/apps/reconstruction/fastMRIreader.py

```diff
@@ -0,0 +1,109 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, Sequence, Tuple, Union
+
+import numpy as np
+from numpy import ndarray
+
+from monai.config import PathLike
+from monai.data.image_reader import ImageReader
+from monai.data.utils import is_supported_format
+from monai.utils import FastMRIKeys, optional_import, require_pkg
+from monai.utils.type_conversion import convert_to_tensor
+
+h5py, has_h5py = optional_import("h5py")
+
+
+@require_pkg(pkg_name="h5py")
+class FastMRIReader(ImageReader):
+    """
+    Load fastMRI files with '.h5' suffix. fastMRI files, when loaded with "h5py",
+    are HDF5 dictionary-like datasets. The keys are:
+
+    - kspace: contains the fully-sampled kspace
+    - reconstruction_rss: contains the root sume of squares of ifft of kspace
+
+    It also has several attributes with the following keys:
+
+    - acquisition (str): acquisition mode of the data
+    - max (float): dynamic range of the data
+    - norm (float): norm of the kspace
+    - patient_id (str): the patient's id whose measurements were recorded
+    """
+
+    def __init__(self):
+        super().__init__()
+
+    def verify_suffix(self, filename: Union[Sequence[PathLike], PathLike]) -> bool:
+        """
+         Verify whether the specified file format is supported by h5py reader.
+
+        Args:
+             filename: file name
+        """
+        suffixes: Sequence[str] = [".h5"]
+        return has_h5py and is_supported_format(filename, suffixes)
+
+    def read(self, data: Union[Sequence[PathLike], PathLike]) -> Dict:  # type: ignore
+        """
+        Read data from specified h5 file.
+        Note that the returned object is a dictionary.
+
+        Args:
+            data: file name to read.
+        """
+        if isinstance(data, (tuple, list)):
+            data = data[0]
+
+        with h5py.File(data, "r") as f:
+            # extract everything from the ht5 file
+            dat = dict(
+                [(key, f[key][()]) for key in f]
+                + [(key, f.attrs[key]) for key in f.attrs]
+                + [("filename", str(data).split("/")[-1])]
+            )
+        f.close()
```

Comment:
Not necessarily, good point.
The only thing is for ipython users. If they call it and f.close() is not there, it'll cause a memory crash for them because ipyhton saves the object in memory if it's not explicitly closed.

Should we care about this use case? If not, I can just remove that. Please let me know what you think, thanks.

Reply:
I think this could be removed, the context manager should already close `f` properly...also, I think `str(data).split("/")[-1]` could be more robust with `os.path.basename(data)`? I'm not familiar with the dataset so I might be wrong